diffrax/integrate.py

```diff
@@ -195,6 +203,168 @@ def _apply_history(x):
+                            _disconts < sub_tnext
+                        )
+                        proxy_candidate = (
+                            _disconts * valid_discont
+                            + jnp.inf * jnp.invert(valid_discont) * _discont
+                        )
+                        best_candidate = jnp.min(proxy_candidate)
+                        best_candidate = jnp.where(
+                            jnp.isinf(best_candidate), sub_tnext, best_candidate
+                        )
+                        discont_update = jnp.where(
+                            jnp.isinf(best_candidate) | (best_candidate == sub_tnext),
+                            False,
+                            True,
+                        )
+                        return best_candidate, discont_update
+
+                def _find_discontinuity_wrapper():
+                    return lax.cond(
+                        jnp.any(init_discont),
+                        _find_discontinuity,
+                        lambda: (tnext, False),
+                    )
+
+                init_discont = jnp.stack(
+                    [jnp.sign(g(sub_tprev)) != jnp.sign(g(sub_tnext)) for g in _gs]
+                )
+                # We might have rejected the step for normal reasons;
+                # skip looking for a discontinuity if so.
+                return lax.cond(
+                    unvmap_any(init_discont),
+                    _find_discontinuity_wrapper,
+                    lambda: (sub_tnext, False),
+                )
+
+            def _always_check_find_discontinuity_wrapper(sub_tprev, sub_tnext):
+                # Only look for a discontinuity on rejected steps.
+                return _maybe_find_discontinuity(sub_tprev, sub_tnext)
+
+            def _maybe_find_discontinuity_wrapper(sub_tprev, sub_tnext):
+                # Only look for a discontinuity on rejected steps.
+                return lax.cond(
+                    keep_step,
+                    lambda _, sub_tnext: (sub_tnext, False),
+                    _maybe_find_discontinuity,
+                    sub_tprev,
+                    sub_tnext,
+                )
+
+            tsearch = jnp.linspace(tprev, tnext, 15)
+            batch_tprev, batch_tnext = tsearch[:-1], tsearch[1:]
+            if delays.recurrent_checking:
+                tnext_candidate, batch_discont_update = jax.vmap(
+                    _always_check_find_discontinuity_wrapper, (0, 0)
+                )(batch_tprev, batch_tnext)
+                proxy_tnext = (
+                    tnext_candidate * batch_discont_update
+                    + jnp.inf * jnp.invert(batch_discont_update) * tnext_candidate
+                )
+                tnext, discont_update = lax.cond(
+                    (proxy_tnext == jnp.inf).all(),
+                    lambda: (tnext, False),
+                    lambda: (jnp.min(proxy_tnext), True),
+                )
```

Comment:
I agree with the first line but not sure the second can be used since `discont_update` is a batched `discont_update` and indicates which `tnext_candidate` are possible. Some of the `tnext_candidate` might not be valid ones.